RavenDB-17760 : Sharding - Storage - Prepare storage to support sharding

aviv86 opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17760

### Additional description

#### use the new voron `DynamicKeyIndexDef` to implement `GetItemsByBucketFrom` for each `ReplicationItem` type :
- documents
- conflicts
- tombstones
- revisions
- attachments
- counters
- timeseries
- deleted timeseries ranges

### Type of change

- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works
 - Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

aviv86 committed:
RavenDB-17760 : change BucketStats tree structure

aviv86 committed:
RavenDB-17760 : update BucketStats on document deletion

aviv86 committed:
RavenDB-17760 : handle conflicts

aviv86 committed:
RavenDB-17760 : avoid possible NRE in SeekForwardFromPrefix if index trees were not created yet

aviv86 committed:
RavenDB-17760 : BucketStats : fix errors, implement GetBucketStats(), add test

aviv86 committed:
RavenDB-17760 : remove all bucket-stats related code

aviv86 committed:
RavenDB-17760 : handle tombstones

aviv86 committed:
RavenDB-17760 : handle revisions

aviv86 committed:
RavenDB-17760 : handle attachments

aviv86 committed:
RavenDB-17760 : handle counters, minor refactoring

aviv86 committed:
RavenDB-17760 : handle timeseries

aviv86 committed:
RavenDB-17760 : handle timeseries deleted ranges

aviv86 committed:
RavenDB-17760 : annotate tests with RavenFact attribute

ayende commented on src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs:
```diff
@@ -2362,6 +2396,24 @@ public IEnumerable<TimeSeriesDeletedRangeItem> GetDeletedRangesFrom(DocumentsOpe
+                foreach (var result in table.SeekForwardFromPrefix(DeleteRangesSchema.DynamicKeyIndexes[DeletedRangesBucketAndEtagSlice], keySlice, prefix, 0))
+                {
+                    var item = CreateDeletedRangeItem(context, ref result.Result.Reader);
+                    if (item.Etag > toEtag)
```
what scenario will we use `toEtag` ?

aviv86 replied:
I added `toEtag` to stay consistent with `GetDeletedRangesFrom(DocumentsOperationContext context, long etag, long toEtag = long.MaxValue)`
but it seems that we never pass `toEtag` to `GetDeletedRangesFrom`

removed `toEtag` from both `GetDeletedRangesFrom` and `GetDeletedRangesByBucketFrom`

ayende commented on src/Voron/Data/BTrees/Tree.cs:
```diff
@@ -260,6 +260,18 @@ private static void ThrowNullReferenceException()
 
+        public void Add(Slice key, byte* value, int size)
+        {
+            if (value == null)
+                ThrowNullReferenceException();
```
Do we need that? We have an assert for this, no?

ayende replied:
Also, why have this entirely? Isn't this just `DirectAdd` ?

aviv86 replied:
not sure why I added this method..
I probably used this earlier in some bucket-stats related code, that I later removed

reverted this

ayende commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -1189,6 +1189,9 @@ public IEnumerable<SeekResult> SeekForwardFrom(TableSchema.AbstractTreeIndexDef
         public IEnumerable<SeekResult> SeekForwardFromPrefix(TableSchema.AbstractTreeIndexDef index, Slice start, Slice prefix, long skip)
         {
             var tree = GetTree(index);
+            if (tree == null)
```
When can this happen?

aviv86 replied:
`tree == null` can happen for example, if we call `GetDocumentsByBucketFrom` when we don't have any documents on this shard yet.

`GetDocumentsByBucketFrom` tries to do 
```
foreach (var result in table.SeekForwardFromPrefix(DocsSchema.DynamicKeyIndexes[DocsBucketAndEtagSlice], keySlice, prefix, 0))
```

but `DocsBucketAndEtagSlice` index-tree doesn't exist yet, it is created in `DocsSchema.Create`, upon first document-put :   
https://github.com/ravendb/ravendb/blob/feature/sharding/src/Raven.Server/Documents/DocumentsStorage.cs#L2483

karmeli87 commented on src/Raven.Server/Documents/ConflictsStorage.cs:
```diff
@@ -706,30 +667,42 @@ public void ThrowConcurrencyExceptionOnConflict(long? expectedEtag, long? curren
-                    if (conflictsTable.Set(tvb))
-                        Interlocked.Increment(ref ConflictsCount);
-                }
+                AddToConflictsTable(incomingChangeVector,
```
Actually it would be great if we can unify all other tables like that as well

aviv86 replied:
added a comment in the issue page to address this when working on bucket-stats

karmeli87 commented on src/Raven.Server/Documents/Sharding/ShardLocator.cs:
```diff
@@ -28,7 +28,7 @@ public static DocumentIdsByShardIndex GroupIdsByShardIndex(IEnumerable<Slice> id
             {
                 var id = ids[i];
-                var shardId = shardedContext.GetShardId(context, id);
+                var shardId = ShardedContext.GetShardId(context, id);
                 var index = shardedContext.GetShardIndex(shardId);
```
let's use the `shardedContext.GetShardIndex(context, id)`

aviv86 replied:
done

aviv86 committed:
RavenDB-17760 : initial work :
- add bucket number to documents table
- add an index on bucket+etag and implement GetDocumentsByBucketFrom
- create buckets-stats tree and update it upon document put/update
- add basic test

aviv86 committed:
RavenDB-17760 : GetCountersByBucketFrom : remove case-sensitive conversion (v4.2 replication-destination is not supported) + cleanup

arekpalinski commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -1359,5 +1390,10 @@ public static IEnumerable<BlittableJsonReaderObject> GetAttachmentsFromDocumentM
             }
         }
+
+        [IndexEntryKeyGenerator]
```
Can we rename `IndexEntryKeyGenerator` to `StorageIndexEntryKeyGenerator` - to make it explicit it's about index storage not ravendb indexes

aviv86 replied:
done

arekpalinski commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2548,6 +2597,32 @@ private IEnumerable<string> IterateCollectionNames(Transaction tx, JsonOperation
+                    yield return TableValueToDocument(context, ref result.Result.Reader);
+                }
+            }
+        }
```
I wonder if we shouldn't introduce general purpose method for getting the items by bucket. Something like `IEnumerable<SeekResult>  GetItemsByBucket(allocator,  table, dynamicindex, bucket, etag)`

Then here and in other places we could have slightly simpler code like:

```
public IEnumerable<Document> GetDocumentsByBucketFrom(DocumentsOperationContext context, int bucket, long etag)
{
    var table = new Table(DocsSchema, context.Transaction.InnerTransaction);

    foreach (var result in  GetItemsByBucket(context.Allocator, table, DocsSchema.DynamicKeyIndexes[DocsBucketAndEtagSlice], bucket, etag)
     {
            yield return TableValueToDocument(context, ref result.Result.Reader);
     }
}
```

aviv86 replied:
done

arekpalinski commented:
Please do the rebase to solve the conflicts

aviv86 committed:
RavenDB-17760 : remove bucket from docs table + cleanup

aviv86 committed:
RavenDB-17760 : docs table + conflicts table : index bucket+etag by using new the voron dynamic-key index, and adjust tests

aviv86 committed:
RavenDB-17760 : fixes after rebase

aviv86 committed:
RavenDB-17760 : address PR comments

ayende commented on src/Raven.Server/Documents/AttachmentsStorage.cs:
```diff
@@ -1359,5 +1385,10 @@ public static IEnumerable<BlittableJsonReaderObject> GetAttachmentsFromDocumentM
         }
+
+        [StorageIndexEntryKeyGenerator]
+        private static ByteStringContext.Scope GenerateBucketAndEtagIndexKeyForAttachments(ByteStringContext context, ref TableValueReader tvr, out Slice slice) =>
+            ExtractIdFromKeyAndGenerateBucketAndEtagIndexKey(context, (int)AttachmentsTable.LowerDocumentIdAndLowerNameAndTypeAndHashAndContentType,
+                (int)AttachmentsTable.Etag, ref tvr, out slice);
```
Let's make this a real function with `{}` - makes it easier to read.

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2548,6 +2592,41 @@ private IEnumerable<string> IterateCollectionNames(Transaction tx, JsonOperation
 
+        public IEnumerable<Document> GetDocumentsByBucketFrom(DocumentsOperationContext context, int bucket, long etag)
+        {
+            var table = new Table(DocsSchema, context.Transaction.InnerTransaction);
```
shouldn't this be `OpenTable()` ?

aviv86 replied:
I don't think we need to use `OpenTable()` here

`table =  new Table(schema, tx)` is meant to be used for global reads only, when we want to use a global index to find data without touching the actual table - which is what we're doing here.

AFAIK `OpenTable(schema, name)` is used for writes, and for reading collection-specific data, e.g. :
 `var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, "Collection.Documents.users"))`

ayende replied:
`OpenTable` implements a cache, so can avoid allocations. The global reads is when we don't have enough information, but here we should have that, no?

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -212,6 +222,16 @@ public IEnumerable<ReplicationBatchItem> GetCountersFrom(DocumentsOperationConte
 
+        public IEnumerable<ReplicationBatchItem> GetCountersByBucketFrom(DocumentsOperationContext context, int bucket, long etag)
+        {
+            var table = new Table(CountersSchema, context.Transaction.InnerTransaction);
```
Shouldn't this be `OpenTable` ?

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -1278,6 +1312,16 @@ public IEnumerable<ReplicationBatchItem> GetTombstonesFrom(DocumentsOperationCon
 
+        public IEnumerable<ReplicationBatchItem> GetTombstonesByBucketFrom(DocumentsOperationContext context, int bucket, long etag)
+        {
+            var table = new Table(TombstonesSchema, context.Transaction.InnerTransaction);
```
Same, and everywhere else

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2548,6 +2592,41 @@ private IEnumerable<string> IterateCollectionNames(Transaction tx, JsonOperation
+            out ByteString buffer)
+        {
+            var scope = allocator.Allocate(sizeof(int) + sizeof(long), out buffer);
+            *(int*)buffer.Ptr = bucket;
```
We don't _have_ to do it, but it would be better if we reversed the bytes here as well.
That would make the sorting of the buckets much more reasonable for us down the line.

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2617,6 +2696,51 @@ public static LazyStringValue TableValueToId(JsonOperationContext context, int i
+            return GenerateBucketAndEtagSlice(context, keyPtr, sizeOfDocId, etagPtr, out slice);
+        }
+
+        private static ByteStringContext.Scope GenerateBucketAndEtagSlice(ByteStringContext context, byte* idPtr, int idSize, byte* etagPtr, out Slice slice)
```
Why are you passing an _etagPtr_ instead of `etag` here?

aviv86 replied:
changed to `etag`

ayende commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -2617,6 +2696,51 @@ public static LazyStringValue TableValueToId(JsonOperationContext context, int i
+            var bucket = ShardHelper.GetBucket(idPtr, idSize);
+
+            *(int*)buffer.Ptr = bucket;
+            *(long*)(buffer.Ptr + sizeof(int)) = *(long*)etagPtr;
```
`*(long*)etagPtr` should be done on the caller, not here.

aviv86 replied:
done

aviv86 committed:
RavenDB-17760 : address PR comments

aviv86 committed:
RavenDB-17760 : fix tests after rebase

aviv86 committed:
RavenDB-17760 : address more PR comments
